### ISSUE TITLE
ICE: TyKind::Error constructed but no error reported, broken MIR in DefId

### ISSUE THREAD

**Issue by chengniansun** [rust-lang/rust#83630]
### Code

```Rust
static DATA: Iterator<Item = _> = "my string";
```

### Meta

`rustc --version --verbose`:
```
rustc 1.53.0-nightly (4a20eb6a9 2021-03-28)
binary: rustc
commit-hash: 4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2
commit-date: 2021-03-28
host: x86_64-unknown-linux-gnu
release: 1.53.0-nightly
LLVM version: 12.0.0
```

### Error output

```
warning: trait objects without an explicit `dyn` are deprecated
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:14
  |
1 | static DATA: Iterator<Item = _> = "my string";
  |              ^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Iterator<Item = _>`
  |
  = note: `#[warn(bare_trait_objects)]` on by default

warning: static is never used: `DATA`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 2 warnings emitted

error: internal compiler error: bad_placeholder_type
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:30
  |
1 | static DATA: Iterator<Item = _> = "my string";
  |                              ^
  |
  = note: delayed at compiler/rustc_typeck/src/collect.rs:370:20

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/compiler/rustc_trait_selection/src/opaque_types.rs:1007:32

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_typeck/src/check/coercion.rs:154:49

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_mir_build/src/build/mod.rs:737:18

error: internal compiler error: mir_const_qualif: MIR had errors
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/transform/mod.rs:237:18

error: internal compiler error: PromoteTemps: MIR had errors
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/transform/promote_consts.rs:55:22

error: internal compiler error: broken MIR in DefId(0:3 ~ perses_node_priority_with_dfs_delta_reduced_mutant[8787]::DATA) ("return type"): bad type [type error]
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:252:27

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:721:20

error: internal compiler error: broken MIR in DefId(0:3 ~ perses_node_priority_with_dfs_delta_reduced_mutant[8787]::DATA) (LocalDecl { mutability: Mut, local_info: None, internal: false, is_block_tail: None, ty: [type error], user_ty: None, source_info: SourceInfo { span: perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1: 1:47 (#0), scope: scope[0] } }): bad type [type error]
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:252:27

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/compiler/rustc_middle/src/ty/relate.rs:334:59

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1014:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (4a20eb6a9 2021-03-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -C opt-level=3 --crate-type staticlib

query stack during panic:
end of query stack
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
warning: trait objects without an explicit `dyn` are deprecated
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:14
  |
1 | static DATA: Iterator<Item = _> = "my string";
  |              ^^^^^^^^^^^^^^^^^^ help: use `dyn`: `dyn Iterator<Item = _>`
  |
  = note: `#[warn(bare_trait_objects)]` on by default

warning: static is never used: `DATA`
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: 2 warnings emitted

error: internal compiler error: bad_placeholder_type
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:30
  |
1 | static DATA: Iterator<Item = _> = "my string";
  |                              ^
  |
  = note: delayed at compiler/rustc_typeck/src/collect.rs:370:20

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/compiler/rustc_trait_selection/src/opaque_types.rs:1007:32

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_typeck/src/check/coercion.rs:154:49

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_mir_build/src/build/mod.rs:737:18

error: internal compiler error: mir_const_qualif: MIR had errors
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/transform/mod.rs:237:18

error: internal compiler error: PromoteTemps: MIR had errors
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/transform/promote_consts.rs:55:22

error: internal compiler error: broken MIR in DefId(0:3 ~ perses_node_priority_with_dfs_delta_reduced_mutant[8787]::DATA) ("return type"): bad type [type error]
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:252:27

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:721:20

error: internal compiler error: broken MIR in DefId(0:3 ~ perses_node_priority_with_dfs_delta_reduced_mutant[8787]::DATA) (LocalDecl { mutability: Mut, local_info: None, internal: false, is_block_tail: None, ty: [type error], user_ty: None, source_info: SourceInfo { span: perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1: 1:47 (#0), scope: scope[0] } }): bad type [type error]
 --> perses_node_priority_with_dfs_delta_reduced_mutant.rs:1:1
  |
1 | static DATA: Iterator<Item = _> = "my string";
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = note: delayed at compiler/rustc_mir/src/borrow_check/type_check/mod.rs:252:27

error: internal compiler error: TyKind::Error constructed but no error reported
  |
  = note: delayed at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/compiler/rustc_middle/src/ty/relate.rs:334:59

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', compiler/rustc_errors/src/lib.rs:1014:13
stack backtrace:
   0: rust_begin_unwind
             at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/library/std/src/panicking.rs:493:5
   1: std::panicking::begin_panic_fmt
             at /rustc/4a20eb6a9da36c88ee929826c4f1eb8d7ea393b2/library/std/src/panicking.rs:435:5
   2: rustc_errors::HandlerInner::flush_delayed
   3: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
   4: core::ptr::drop_in_place<rustc_session::parse::ParseSess>
   5: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
   6: core::ptr::drop_in_place<rustc_interface::interface::Compiler>
   7: rustc_span::with_source_map
   8: rustc_interface::interface::create_compiler_and_run
   9: scoped_tls::ScopedKey<T>::set
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.53.0-nightly (4a20eb6a9 2021-03-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -C opt-level=3 --crate-type staticlib

query stack during panic:
end of query stack
```

</p>
</details>

**Comment by chengniansun** [comment]
This might be a duplicate of #83628, though this bug requires the flag `-C opt-level=3`.

**Comment by jyn514** [comment]
> This might be a duplicate of #83628, though this bug requires the flag `-C opt-level=3`.

I can reproduce without -C opt-level=3, although the error is different:

```

Global is external, but doesn't have external or weak linkage!
{}* @_ZN10playground4DATA17h640999b0687ef15dE
LLVM ERROR: Broken module found, compilation aborted!
```

**Comment by jyn514** [comment]
If I add `pub`, I get the ICEs in debug mode as well: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=ff07815db1102032bf1dc1e67f20ebff

**Comment by apiraino** [comment]
Assigning priority as discussed as part of the [Prioritization Working Group procedure](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.2383630.20ICE.3A.20TyKind.3A.3AError.20constructed.20but.20no.20error.20reported.E2.80.A6) and removing `I-prioritize`.

@rustbot label -I-prioritize +P-medium

**Comment by JohnTitor** [comment]
Fixed by #83739.
